Overhaul Apple disk implementation

complexspaces opened the pull request:
This PR rewrites a significant portion of the Apple disk handling code, partially in response to https://github.com/GuillaumeGomez/sysinfo/issues/841. Some of the notable improvements are:
- Better detection of devices that are actually ejectable, but just don’t report as such.
- Continues to skip most[^1] of the hidden APFS partitions on modern Apple devices.
- More correctly determines the name of external drives. In my test case, it detected an external SSD had the name External” and not “Untitled 1”, which matches disk utility.
- More accurate available capacity numbers
- Almost[^2] entirely supported on iOS (!)

In more detail, this PR completely replaces `DiskArbitration.framework` with `CoreFoundation`'s volume handling instead. This was done for three reasons:
1. `CoreFoundation` allows accessing the "visibility" property of a volume/disk. This is very helpful for filtering out internal partitions and snapshots.
2. It is fully supported on all supported versions of iOS, unlike `DiskArbitration.framework` which is solely available on macOS.
3. Its supported in the default macOS app sandbox.

By utilizing `kCFURLVolumeIsBrowsableKey`, a volume property that tells you if its browsable on the system by default, which matches other Apple apps such as Finder volumes, disk utility, and system information's storage devices list. Using this framework also allows us to access better space estimation values that match what the rest of the system would report.

In addition, I took the chance to remove the "hacky" implementation of disk medium type detection and replace it with something well-documented via IOKit. This change is also compatible with the default macOS sandbox, further reducing the feature gap of the crate with and without the `apple-sandbox` feature enabled.

Finally, I removed the crate's build script to improve compilation times. Ever since Objective-C code was removed from the crate, it hasn't been strictly needed. The `#[link]` attribute can do the same thing the current version of it was doing to make sure `IOKit` functions and constants are available.

This PR was tested in and outside of the default macOS app sandbox with a variety of external drives attached, and inside the iOS simulator and on a real iOS device.

Closes https://github.com/GuillaumeGomez/sysinfo/issues/841

[^1]: In order to make some use cases possible, the disk list now returns `/System/Volumes/Data` too.
[^2]: Disk type detection is currently unsupported on iOS due to version requirements.

GuillaumeGomez commented on build.rs:
```diff
@@ -1,19 +0,0 @@
-fn main() {
```
The removal of this file makes me really happy.

GuillaumeGomez commented on src/apple/disk.rs:
```diff
@@ -53,14 +62,330 @@ impl DiskExt for Disk {
+            return Vec::new();
+        }
+
+        disks.set_len(count as usize);
```
```suggestion
        disks.set_len(count as _);
```

GuillaumeGomez replied:
I just had to fix clippy lints related to unneeded casts because a type changed because of a platform change that I'd much rather prefer to have a generic cast.

complexspaces replied:
This one isn't platform specific, it comes out of the Rust standard library. It would be a breaking change if they made this not take a `usize` so I am not sure that I have the same level of worry with this one.

GuillaumeGomez replied:
Oh indeed. Sorry, I don't why I was sure it was another type. My bad. Resolved then.

GuillaumeGomez commented on src/apple/disk.rs:
```diff
@@ -53,14 +62,330 @@ impl DiskExt for Disk {
+                false as u8,
+            )
+        })
+        .unwrap();
```
Please use a `match` and add a `sysinfo_debug!` call in case it failed. I'm really uneasy to just `unwrap`.

GuillaumeGomez commented on src/apple/disk.rs:
```diff
@@ -53,14 +62,330 @@ impl DiskExt for Disk {
+    }
+}
+
+pub(super) fn get_disks() -> Vec<Disk> {
```
I think this whole function could be turned unsafe. There is no point to have so many inner `unsafe` blocks.

complexspaces replied:
I made an inner function that's unsafe solely for the purpose of removing these blocks. `get_disks()` now just calls it. I didn't want to make the outer-most one unsafe because it bled implementation details into `system.rs` unnecessarily.

GuillaumeGomez replied:
There is no point in that. If `system.rs` needs an `unsafe` block, then so be it. In this case it just creates an unneeded indirection.

complexspaces replied:
I'll do that if its your preference then, even though I don't necessarily agree this is how `unsafe` should be used.

GuillaumeGomez commented on src/apple/disk.rs:
```diff
@@ -53,14 +62,330 @@ impl DiskExt for Disk {
+        // Future note: There is a difference between `kCFURLVolumeIsBrowsableKey` and the
+        // `kCFURLEnumeratorSkipInvisibles` option of `CFURLEnumeratorOptions`. Specifically,
+        // the first one considers the writable `Data`(`/System/Volumes/Data`) partition to be
+        // browsable, while it is classified as "invisible" by CoreFoundation's volume emumerator.
```
Nice comment. :+1:

GuillaumeGomez commented on src/apple/disk.rs:
```diff
@@ -53,14 +62,330 @@ impl DiskExt for Disk {
+            )
+        })
+    }
+    .unwrap() as u64
```
Same, please don't `unwrap`.

GuillaumeGomez commented on src/apple/disk.rs:
```diff
@@ -53,14 +62,330 @@ impl DiskExt for Disk {
+            _defined = CFReleaser::new(cfs::CFStringCreateWithBytesNoCopy(
+                kCFAllocatorDefault,
+                val.as_ptr(),
+                val.len() as isize,
```
`as _`

GuillaumeGomez replied:
It's mostly because the `long` size might change depending on the arch. Like mentioned in a previous comment, I had to fix clippy lints about this recently that I'd rather get some "head start" on this.

complexspaces replied:
This is using `as _` now to avoid any issues that might show up later 👍

GuillaumeGomez commented on src/apple/disk.rs:
```diff
@@ -53,14 +62,330 @@ impl DiskExt for Disk {
+                false as _,
+                core_foundation_sys::base::kCFAllocatorNull,
+            ))
+            .unwrap();
```
Same: no `unwrap` please.

GuillaumeGomez commented on src/apple/disk.rs:
```diff
@@ -53,14 +62,330 @@ impl DiskExt for Disk {
+    mount_point: PathBuf,
+    volume_url: RetainedCFURL,
+    c_disk: libc::statfs,
+    disk_props: &CFReleaser<core_foundation_sys::dictionary::__CFDictionary>,
```
Don't you have `RetainedCFDictionary` for this one?

GuillaumeGomez commented on src/apple/disk.rs:
```diff
@@ -53,14 +62,330 @@ impl DiskExt for Disk {
+        )
+    }
+    .map(OsString::from)
+    .unwrap();
```
Use `?` instead.

GuillaumeGomez commented on src/apple/disk.rs:
```diff
@@ -53,14 +62,330 @@ impl DiskExt for Disk {
+            disk_props.inner(),
+            DictKey::Extern(ffi::kCFURLVolumeIsEjectableKey),
+        )
+        .unwrap();
```
Use `?` instead.

GuillaumeGomez commented on src/apple/disk.rs:
```diff
@@ -53,14 +62,330 @@ impl DiskExt for Disk {
+            disk_props.inner(),
+            DictKey::Extern(ffi::kCFURLVolumeIsRemovableKey),
+        )
+        .unwrap();
```
Use `?` instead.

GuillaumeGomez commented on src/apple/disk.rs:
```diff
@@ -53,14 +62,330 @@ impl DiskExt for Disk {
+                disk_props.inner(),
+                DictKey::Extern(ffi::kCFURLVolumeIsInternalKey),
+            )
+            .unwrap();
```
Use `?` instead.

GuillaumeGomez commented on src/apple/disk.rs:
```diff
@@ -53,14 +62,330 @@ impl DiskExt for Disk {
+    })
+}
+
+fn new_disk(
```
Same: there are too many `unsafe` blocks to not just turn this function into an unsafe one.

GuillaumeGomez commented on src/apple/ffi.rs:
```diff
@@ -1,21 +1,31 @@
-pub struct __DADisk(c_void);
-#[repr(C)]
-pub struct __DASession(c_void);
+#[link(name = "CoreFoundation", kind = "framework")]
```
Thank you again for removing the `build.rs` script!

complexspaces replied:
Its always nice to see another compilation requirement go away 😄

GuillaumeGomez commented on src/apple/macos/disk.rs:
```diff
@@ -1,200 +1,120 @@
+            kCFAllocatorNull,
+        )
+    })
+    .unwrap();
```
Please return an `Option` instead (and use `sysinfo_debug!` if it's `None`).

GuillaumeGomez commented on src/apple/macos/disk.rs:
```diff
@@ -1,200 +1,120 @@
+        CStr::from_ptr(disk.f_mntfromname.as_ptr())
+            .to_bytes()
+            .strip_prefix(b"/dev/")
+            .expect("device mount point in unknown format")
```
No failure, please return `None` in this case and add a debug log with `sysinfo_debug!`.

GuillaumeGomez commented on src/apple/macos/disk.rs:
```diff
@@ -1,200 +1,120 @@
+            .expect("device mount point in unknown format")
+    };
+
+    let matching =
```
Can you add a comment about this saying that there is no need to wrap it into a `CFReleaser` because `IOServiceGetMatchingServices` will take ownership of it please?

GuillaumeGomez commented on src/apple/macos/disk.rs:
```diff
@@ -1,200 +1,120 @@
-                utils::vec_to_rust(buf)
-            } else {
-                None
+    let service_iterator = IOReleaser::new(service_iterator).unwrap();
```
No `unwrap` please.

GuillaumeGomez commented on src/apple/macos/disk.rs:
```diff
@@ -1,200 +1,120 @@
+        IOReleaser::new(unsafe { ffi::IOIteratorNext(service_iterator.inner()) })
+    {
+        loop {
+            let result = unsafe {
```
What's the point of storing into `result` since it's only used for a `if` comparison?

GuillaumeGomez commented:
It looks like a great start. Mostly nits at this point. Thanks for working on this!

GuillaumeGomez commented:
Hi @complexspaces. Do you plan to continue working on this PR?

complexspaces commented:
Hey there @GuillaumeGomez. I do plan on working on this again, sorry. I have been busy enough with work that this one got de-prioritized. I'll try and have it cleaned up by the weekend at the latest.

complexspaces commented on src/apple/disk.rs:
```diff
@@ -53,14 +62,330 @@ impl DiskExt for Disk {
+type RetainedCFDictionary = CFReleaser<core_foundation_sys::dictionary::__CFDictionary>;
+type RetainedCFURL = CFReleaser<core_foundation_sys::url::__CFURL>;
+
+unsafe fn build_requested_properties(properties: &[CFStringRef]) -> RetainedCFArray {
```
Done 👍. The callers now log and skip their action if this allocation fails.

complexspaces commented:
@GuillaumeGomez This is ready for another round of review. I apologize for taking so long to finish this up.

Outside of addressing the review comments, the most notable change is https://github.com/GuillaumeGomez/sysinfo/pull/855/commits/c323fae3d96148e1a186f71a2a2ad6c1ef767697, where I improved on both of the Apple auto-release objects. I've encoded their invariants in the type system and added to the API to make them match better with some infallible use cases among the new code.

GuillaumeGomez commented on src/apple/disk.rs:
```diff
@@ -53,14 +62,338 @@ impl DiskExt for Disk {
+        let volume_url = match CFReleaser::new(
+            core_foundation_sys::url::CFURLCreateFromFileSystemRepresentation(
+                kCFAllocatorDefault,
+                c_disk.f_mntonname.as_ptr() as *const u8,
```
```suggestion
                c_disk.f_mntonname.as_ptr() as *const _,
```

GuillaumeGomez commented on src/apple/disk.rs:
```diff
@@ -53,14 +62,338 @@ impl DiskExt for Disk {
+        let volume_url = match CFReleaser::new(
+            core_foundation_sys::url::CFURLCreateFromFileSystemRepresentation(
+                kCFAllocatorDefault,
+                c_disk.f_mntonname.as_ptr() as *const u8,
+                c_disk.f_mntonname.len() as isize,
+                false as u8,
```
```suggestion
                c_disk.f_mntonname.as_ptr() as *const _,
                c_disk.f_mntonname.len() as _,
                false as _,
```

complexspaces replied:
Done 👍

GuillaumeGomez commented on src/apple/disk.rs:
```diff
@@ -53,14 +62,338 @@ impl DiskExt for Disk {
+        )
+        .unwrap_or_default();
+
+        // Do not return invisible "disks." Most of the time, these are APFS snapshots, hidden
```
Little typo:

```suggestion
        // Do not return invisible "disks". Most of the time, these are APFS snapshots, hidden
```

complexspaces replied:
Good catch, thank you.

GuillaumeGomez commented on src/apple/disk.rs:
```diff
@@ -53,14 +62,338 @@ impl DiskExt for Disk {
+fn get_available_volume_space(disk_props: &RetainedCFDictionary) -> u64 {
+    // We prefer `AvailableCapacityForImportantUsage` over `AvailableCapacity` because
+    // it takes more of the system's properties into account, like the trash, system-managed caches,
+    // etc. It generally also returns higher values too, because of the above, so its a more accurate
```
Little typo:

```suggestion
    // etc. It generally also returns higher values too, because of the above, so it's a more accurate
```

GuillaumeGomez commented on src/apple/disk.rs:
```diff
@@ -53,14 +62,338 @@ impl DiskExt for Disk {
+    get_dict_value(dict, key, |v| {
+        let v = v as cfs::CFStringRef;
+
+        let len_utf16 = cfs::CFStringGetLength(v);
+        let len_bytes = len_utf16 as usize * 2; // Two bytes per UTF-16 codepoint.
```
```suggestion
        let len_utf16: usize = cfs::CFStringGetLength(v) as _;
        let len_bytes = len_utf16 * 2; // Two bytes per UTF-16 codepoint.
```

complexspaces replied:
I've cleaned this one up too. It now does `as usize` when binding `len_utf16`.

GuillaumeGomez commented on src/apple/disk.rs:
```diff
@@ -53,14 +62,330 @@ impl DiskExt for Disk {
+        if CFNumberGetValue(
+            v.cast(),
+            core_foundation_sys::number::kCFNumberSInt64Type,
+            &mut val as *mut i64 as *mut c_void,
```
It comes from two things:
 * The function header might change (unlikely though but still something I'm wary of).
 * Coherency with the rest.

A very small point in favour of generic casting too: it's a bit shorter (no need to import `c_void` too).

But in any case, if you don't want to change it, I won't be too annoying for this one as it is very unlikely it'll ever change. Mostly a personal preference at this point.

complexspaces replied:
I think that this one is worth keeping as-is for the following reasons:
- We already need to import `c_void` for other functions in this file, so we don't save an import
- CoreFoundation objects are always typed as `c_void` throughout the headers and it would be a huge and unlikely change to move away from that. 
- This function also needs to take an arbitrary "number-like" CF value so it makes sense that it needs to be type erased regardless
- `val` will always be an `i64` since we declare the value here in Rust.

GuillaumeGomez replied:
Ok, makes sense. :+1:

GuillaumeGomez commented on src/apple/macos/disk.rs:
```diff
@@ -1,200 +1,125 @@
+        CFReleaser::new_unchecked(cfs::CFStringCreateWithBytesNoCopy(
+            kCFAllocatorDefault,
+            ffi::kIOPropertyDeviceCharacteristicsKey.as_ptr(),
+            ffi::kIOPropertyDeviceCharacteristicsKey.len() as CFIndex,
```
```suggestion
            ffi::kIOPropertyDeviceCharacteristicsKey.len() as _,
```

complexspaces replied:
Fixed too 👍

GuillaumeGomez commented on src/apple/macos/disk.rs:
```diff
@@ -1,200 +1,125 @@
+    while let Some(mut current_service_entry) =
+        IOReleaser::new(unsafe { ffi::IOIteratorNext(service_iterator.inner()) })
+    {
+        loop {
```
It'll always either `break` (to continue le `while let`) or `return`. I think this `loop` can be removed then, no?

complexspaces replied:
This loop still needs to exist AFAIK. Without this loop in place, we don't recurse far enough through the IO registry entry's properties to find the disk information we're looking for. For example if I remove this, the function starts becoming unable to get the correct disk information since after a single call to `IORegistryEntryCreateCFProperty` returns nothing, it quits.

complexspaces replied:
Would it help if I added a comment to this effect?

GuillaumeGomez replied:
I think I got it but a comment would indeed be very welcome. :+1:

complexspaces replied:
I've added a comment about this to the loop so that those looking at the code in the future have more context 👍.

GuillaumeGomez commented on src/apple/macos/ffi.rs:
```diff
@@ -38,89 +43,57 @@ extern "C" {
+        #[allow(dead_code)]
+        pub fn IOConnectCallStructMethod(
+            connection: mach_port_t,
+            selector: u32,
```
This is wrong here. From the [documentation](https://developer.apple.com/documentation/swift/int):

> On 32-bit platforms, Int is the same size as Int32, and on 64-bit platforms, Int is the same size as Int64.

Better make a type alias for it.

GuillaumeGomez replied:
Sorry, wrong line, it's for `inputStructCnt: size_t,` just below.

complexspaces replied:
I'm not sure that I follow here, sorry. [The docs](https://developer.apple.com/documentation/iokit/1514274-ioconnectcallstructmethod?language=objc) for `IOConnectCallStructMethod` say that `inputStructCnt` is a `size_t` in the header. The current definition here in `sysinfo` is, as you noted, `inputStructCnt: size_t`, where `size_t` is the `libc` import. Isn't that already correct?

GuillaumeGomez replied:
Sorry, I was looking at the wrong docs once again... This one to be exact: https://developer.apple.com/documentation/iokit/1514274-ioconnectcallstructmethod

GuillaumeGomez commented on src/apple/macos/ffi.rs:
```diff
@@ -38,89 +43,57 @@ extern "C" {
+            inputStructCnt: size_t,
+            outputStruct: *mut KeyData_t,
+            outputStructCnt: *mut size_t,
+        ) -> i32;
```
```suggestion
        ) -> kern_return_t;
```

complexspaces replied:
Fixed.

GuillaumeGomez commented on src/apple/macos/ffi.rs:
```diff
@@ -38,89 +43,57 @@ extern "C" {
+
+        pub fn IOServiceOpen(device: io_object_t, a: u32, t: u32, x: *mut io_connect_t) -> i32;
+
+        pub fn IOServiceClose(a: io_connect_t) -> i32;
```
```suggestion
        pub fn IOServiceClose(a: io_connect_t) -> kern_return_t;
```

complexspaces replied:
Fixed as well.

GuillaumeGomez commented on src/apple/macos/ffi.rs:
```diff
@@ -38,89 +43,57 @@ extern "C" {
+    extern "C" {
+        pub fn IOServiceMatching(a: *const c_char) -> *mut c_void;
+
+        pub fn IOServiceOpen(device: io_object_t, a: u32, t: u32, x: *mut io_connect_t) -> i32;
```
```suggestion
        pub fn IOServiceOpen(
            device: io_service_t, // please create a type alias `pub type io_service_t = io_object_t;`
            owning_task: task_port_t,
            type_: u32,
            connect: *mut io_connect_t,
        ) -> kern_return_t;
```

complexspaces replied:
This signature binding has been cleaned up now 👍

GuillaumeGomez commented on src/apple/macos/ffi.rs:
```diff
@@ -38,89 +43,57 @@ extern "C" {
-    pub type io_iterator_t = io_object_t;
+
+    extern "C" {
+        pub fn IOServiceMatching(a: *const c_char) -> *mut c_void;
```
```suggestion
        pub fn IOServiceMatching(name: *const cchar) -> CFDictionary;
```

[cchar](https://developer.apple.com/documentation/swift/cchar) is **not** a `c_char` but a `i8`. Please make a type alias for it.

GuillaumeGomez replied:
Also I think it's `CFMutableDictionaryRef` to be exact.

complexspaces replied:
Regarding the `cchar` difference, I think that you might have looked at the wrong set of Apple documentation. `cchar` is a Swift type but we're not using any Swift bindings here. If you look at the [Objective-C docs](https://developer.apple.com/documentation/iokit/1514687-ioservicematching?language=objc) for this function, it lists the type as `const char *name`.

I'll switch in the more-correct return type though 👍

GuillaumeGomez replied:
Oh you're right. Sorry about that...

complexspaces replied:
No worries at all. I've switched this to correctly return a `CFMutableDictionaryRef` as well.

GuillaumeGomez commented on src/apple/macos/ffi.rs:
```diff
@@ -38,89 +43,57 @@ extern "C" {
+    pub fn IOBSDNameMatching(
+        mainPort: mach_port_t,
+        options: u32,
+        bsdName: *const c_char,
```
```suggestion
        bsdName: *const cchar,
```

Again, `cchar` isn't the same as `char`.

complexspaces replied:
Resolving per the conversation about about Swift vs Objective-C types.

GuillaumeGomez commented on src/apple/disk.rs:
```diff
@@ -53,14 +62,338 @@ impl DiskExt for Disk {
+        DictKey::Extern(val) => val,
+        #[cfg(target_os = "macos")]
+        DictKey::Defined(val) => {
+            _defined = CFReleaser::new_unchecked(cfs::CFStringCreateWithBytesNoCopy(
```
Please add a comment explaining why it's safe to get this pointer without checking it.

complexspaces replied:
My original idea didn't work out here so I've removed this in favor of the variant that just does a `NULL` check. Hopefully that is simpler to maintain going forward as well.

GuillaumeGomez commented on src/utils.rs:
```diff
@@ -3,7 +3,7 @@
 #[cfg(all(
     not(feature = "unknown-ci"),
-    any(target_os = "linux", target_os = "android", target_vendor = "apple")
+    any(target_os = "linux", target_os = "android")
```
As a follow-up once this PR is merged, we can move this function into the `linux` folder directly since it's only used by linux targets.

GuillaumeGomez replied:
Done in https://github.com/GuillaumeGomez/sysinfo/pull/873.

complexspaces replied:
Thanks for handling that!

GuillaumeGomez replied:
:+1:

complexspaces commented on src/apple/macos/disk.rs:
```diff
@@ -1,200 +1,120 @@
+        cfs::CFStringCreateWithBytesNoCopy(
+            kCFAllocatorDefault,
+            ffi::kIOPropertyDeviceCharacteristicsKey.as_ptr(),
+            ffi::kIOPropertyDeviceCharacteristicsKey.len() as isize,
```
This one is now using `as _` as well.

GuillaumeGomez commented:
One comment to resolve and one comment to add and it'll be ready to go. I'm so excited about these changes to be merged. :)

GuillaumeGomez commented:
Thanks a lot for this work!

complexspaces commented:
Thank you for working through this with me, I'm glad to see it merged 🎉
